File: api/fetch-info.js
// File: /api/fetch-info.js (Using @distube/ytdl-core)
const ytdl = require('@distube/ytdl-core');

// Helper function to format duration from seconds to HH:MM:SS
const formatDuration = (seconds) => {
    if (isNaN(seconds) || seconds === null) return 'N/A';
    return new Date(parseInt(seconds) * 1000).toISOString().substr(11, 8);
};

module.exports = async (req, res) => {
    // Standard CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    const { url } = req.query;

    if (!url) {
        return res.status(400).json({ success: false, error: 'Missing YouTube URL.' });
    }

    try {
        const info = await ytdl.getInfo(url);
        const details = info.videoDetails;

        const responseData = {
            type: 'video',
            id: details.videoId,
            title: details.title,
            thumbnail: details.thumbnails[details.thumbnails.length - 1].url,
            duration: details.isLive ? 'Live' : formatDuration(details.lengthSeconds),
        };
        
        return res.status(200).json({ success: true, data: responseData });

    } catch (error) {
        console.error('--- DETAILED FETCH ERROR ---');
        console.error(error);
        console.error('--- END DETAILED ERROR ---');
        
        res.status(500).json({ 
            success: false, 
            error: `Failed to get video info: ${error.message}` 
        });
    }
};

File: api/fetch-info.py
# File: api/fetch-info.py
from flask import Flask, request, jsonify
from yt_dlp import YoutubeDL
import time

app = Flask(__name__)

# Helper function to format seconds into HH:MM:SS
def format_duration(seconds):
    if seconds is None:
        return 'N/A'
    return time.strftime('%H:%M:%S', time.gmtime(seconds))

@app.route('/api/fetch-info', methods=['GET'])
def fetch_info():
    url = request.args.get('url')
    if not url:
        return jsonify({"success": False, "error": "Missing YouTube URL."}), 400

    try:
        # Use yt-dlp to extract information
        with YoutubeDL({'quiet': True}) as ydl:
            info_dict = ydl.extract_info(url, download=False)
            
            response_data = {
                "type": "video",
                "id": info_dict.get("id"),
                "title": info_dict.get("title"),
                "thumbnail": info_dict.get("thumbnail"),
                "duration": format_duration(info_dict.get("duration")),
            }
            
            return jsonify({"success": True, "data": response_data})

    except Exception as e:
        print(f"--- DETAILED FETCH ERROR --- \n{e}\n--- END ERROR ---")
        return jsonify({"success": False, "error": f"Failed to get video info: {str(e)}"}), 500

# This is the entry point for Vercel
def handler(request, context):
    return app(request, context)